### PR TITLE
feat(router): add ability to recover from malformed url

### PR DESCRIPTION
**Pull request by skryvets**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Using invalid character sequences (like %) in the query param throws an exception that can't be catched anywhere and causing the app to crash.

Issue Number: [21468](https://github.com/angular/angular/issues/21468)

## What is the new behavior?
Code that might throw an exception is wrapped inside the `try/catch` construction preventing app from crash.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -490,15 +510,23 @@ export class Router {
...
+  }
 
   /** Returns whether the url is activated */
   isActive(url: string|UrlTree, exact: boolean): boolean {
     if (url instanceof UrlTree) {
       return containsTree(this.currentUrlTree, url, exact);
     }
 
-    const urlTree = this.urlSerializer.parse(url);
+    const urlTree = this.parseUrl(url);
```
Comment (sergeome) [inline packages/router/src/router.ts]: I changed this to use internal method `parseUrl()` which has `try/catch` construction when parsing the URL. Is that qualified as a breaking change?

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -160,6 +160,16 @@ function defaultErrorHandler(error: any): any {
   throw error;
 }
 
+/**
+ * @whatItDoes Malformed uri error handler is invoked when `encodedURI` contains invalid character
+ * sequences.
+ */
+export type MalformedUriErrorHandler = (urlSerializer: UrlSerializer, url: string) => UrlTree;
+
+function defaultMalformedUriErrorHandler(urlSerializer: UrlSerializer, url: string): UrlTree {
```
Comment (sergeome) [inline packages/router/src/router.ts]: The `defaultMalformedUriErrorHandler` is redirecting to the root url in a case of an error. In order to comply with type `MalformedUriErrorHandler` it should have two params `urlSerializer` and `url`. 
This is an example of the method call:
```
this.malformedUriErrorHandler(this.urlSerializer, '/');
```

However, I would like to avoid passing root url from outside. It seems more expected when the default behavior is defined inside the method:
```
function defaultMalformedUriErrorHandler(urlSerializer: UrlSerializer, url: string): UrlTree {
  return urlSerializer.parse('/');
}
```

If I would do that, I will have "unnecessary" url param (since it's not used by defaultHandler), but it's needed to stay to comply with method type.

**Question**. What is better:
1. Pass root url from outside.
2. Implement it as an internal part by not using the `url` param.

**Comment by sergeome** [comment]
I implemented this fix, but have few doubts (commented in the code).

File: packages/router/src/router.ts (hunk 2 above)
Comment (wKoza) [reply packages/router/src/router.ts]: Why do you want to add a param `url` in your type `MalformedUriErrorHandler` ? Is not it the role of the implementation to define the path ?

Comment (sergeome) [reply packages/router/src/router.ts]: So the user can configure it by providing their own handling function. Please see [this part](https://github.com/angular/angular/pull/23283/files#diff-ff3b9c3c80dbf5eb828cf6d919746075R303).

Comment (wKoza) [reply packages/router/src/router.ts]: Ok, but in this case, we could simply have a your url in `ExtraOptions` like this:
```javascript
interface ExtraOptions { 
  pathWhenMalformedUriError?: string // Sorry for the name 
```

and change your function by:
```javascript
function malformedUriErrorHandler(urlSerializer: UrlSerializer, url: string): UrlTree {
  return urlSerializer.parse(url);
}
```

And your catch by:
```javascript
urlTree = this.malformedUriErrorHandler(this.urlSerializer, this.pathWhenMalformedUriError);
```

Of course, Router has the default value '\'.

Comment (sergeome) [reply packages/router/src/router.ts]: @jasonaden,

Could you please share you thoughts on this? Thank you!

**Review by jasonaden — APPROVED** [review]
Overall this looks good. I made a couple minor adjustments (you'll see three `fixup!` commits which will get squashed when this is merged). I also ran the tools to update the golden public API file since this adds to the public APIs.

Overall this is great and will be quite helpful to the community. Thanks!

File: packages/router/src/router.ts [hunk 3]
```diff
@@ -160,6 +160,19 @@ function defaultErrorHandler(error: any): any {
   throw error;
 }
 
+/**
+ * @whatItDoes Malformed uri error handler is invoked when `Router.parseUrl(url)` throws an
```
Comment (IgorMinar) [inline packages/router/src/router.ts]: this is an old jsdoc tag, we no longer use it. please use @description instead. The first sentence should be descriptive and meaningful by itself. See other @description tags in this file for examples.

File: tools/public_api_guard/router/router.d.ts [hunk 4]
```diff
@@ -339,6 +340,7 @@ export declare class Router {
     config: Routes;
     errorHandler: ErrorHandler;
     readonly events: Observable<Event>;
+    malformedUriErrorHandler: MalformedUriErrorHandler;
```
Comment (IgorMinar) [inline tools/public_api_guard/router/router.d.ts]: maybe I don't understand what's going on here, but I find it weird that the custom url handler is exposed on the router class. Why is that needed? shouldn't the router call this handler when it needs it?

I'd expect the router to be provided only via router configuration and not exposed anywhere else.

**Review by IgorMinar — DISMISSED** [review]

**Comment by wKoza** [comment]
@jasonaden , I don't understand why we should have a url parameter on the type `MalformedUriErrorHandler`. We say:
> `'url'` -  url redirect to in a case of an error.

But, it's wrong. We should say:
> `'url'` -  default url redirect to provided by Angular in a case of an error 

`url` does not have much interest, it is always valued to `\`:

```javascript
} catch (e) {
      urlTree = this.malformedUriErrorHandler(e, this.urlSerializer, '/');
}
```

**Comment by jasonaden** [comment]
@wKoza So the reason for passing the `url` here is that it's the url that failed to parse. Maybe as an application developer you want to recover in a different way than just going to `/`. Maybe you want to detect things such as non-encoded `%` sign, determine it's not part of a percent encoding set, and encode it manually. For example:

Incoming URL that will fail: `http://mysite.com/foo%bar`

Possible recovery URL: `http://mysite.com/foo%25bar`

You can only make this type of recovery if we pass the offending URL to the error handler.

File: tools/public_api_guard/router/router.d.ts (hunk 4 above)
Comment (jasonaden) [reply tools/public_api_guard/router/router.d.ts]: Marking as `@internal` and inlining in `ExtraOptions` after discussing with Igor.

**Comment by wKoza** [comment]
thanks @jasonaden. Indeed, that's what we can expect. But we should not do this instead:

```javascript
parseUrl(url: string): UrlTree {
    let urlTree: UrlTree;
    try {
      urlTree = this.urlSerializer.parse(url);
    } catch (e) {
      urlTree = this.malformedUriErrorHandler(e, this.urlSerializer, url);
    }
    return urlTree;
 }
```

If we want to catch the url ?

**Comment by jasonaden** [comment]
@wKoza Yes, you're right. My push back to this branch appears to have gotten blocked last week, but see the code now. It's what you have noted above, plus the changes @IgorMinar wanted.

File: packages/router/src/router_module.ts [hunk 5]
```diff
@@ -300,6 +300,18 @@ export interface ExtraOptions {
...
   paramsInheritanceStrategy?: 'emptyOnly'|'always';
+
+  /**
+   * A custom malformed uri error handler function. This handler is invoked when encodedURI contains
+   * invalid character sequences. The default implementation is to redirect to the root url dropping
+   * any path or param info. This function passes three parameters:
+   *
+   * - `'URIError'` - Error thrown when parsing a bad URL
+   * - `'UrlSerializer'` - UrlSerializer that’s configured with the router.
+   * - `'url'` -  url redirect to in a case of an error.
```
Comment (wKoza) [inline packages/router/src/router_module.ts]: @jasonaden , `url`definition should be `Url causing the error`

**Comment by jasonaden** [comment]
I couldn't mark `Router.malformedUriErrorHandler` as `@internal` because it needs to be used in the `RouterTestingModule`. So it's a public property, but the type info is inlined so it will behave correctly in documentation.

**Comment by wbhob** [comment]
@sergeome what's the status of this PR? Looks like it hasn't been touched for a few months. CodeTriage sent me here to check in.

**Comment by sergeome** [comment]
@wbhob, as far as I understand there some internal discussions should happen within the Angular team.
@jasonaden probably can speak to that. 
Please, correct me if I'm wrong.

**Review by mhevery — APPROVED** [review]
